mfsmanpages/mfsgui.8
.TH mfsgui "8" "May 2026" "MooseFS 4.59.2-1" "This is part of MooseFS"
.SH NAME
mfsgui \- start, restart or stop MooseFS GUI server
.SH SYNOPSIS
.B mfsgui
[\fB\-c\fP \fICFGFILE\fP]
[\fB\-f\fP]
[\fB\-d\fP]
[\fB\-u\fP]
[\fB\-n\fP]
[\fB\-t\fP \fILOCKTIMEOUT\fP]
[\fIACTION\fP]
.PP
.B mfsgui \-v
.PP
.B mfsgui \-h
.SH DESCRIPTION
.PP
\fBmfsgui\fP is the MooseFS GUI server. It consist of a HTTP server and scripts that make up the GUI interface to MooseFS. The user can access the GUI via any web browser, using the IP (or appropriate domain name) and port defined in \fBmfsgui.cfg\fP.
.SH OPTIONS
.TP
\fB\-v\fP
print version information and exit
.TP
\fB\-h\fP
print usage information and exit
.TP
\fB\-c\fP \fICFGFILE\fP
specify alternative path of configuration file (default is
\fBmfsgui.cfg\fP in system configuration directory)
.TP
\fB\-f\fP
run in foreground, don't daemonize
.TP
\fB\-d\fP
run with dangerous options (names: DANGEROUS_*)
.TP
\fB\-u\fP
log undefined configuration values (when default is assumed)
.TP
\fB\-n\fP
do not attempt to increase limit of core dump size
.TP
\fB\-t\fP \fILOCKTIMEOUT\fP
how long to wait for lockfile (in seconds; default is 60 seconds)
.TP
\fIACTION\fP
tells MooseFS GUI server to perform one of the following actions:
.RS 
.IP start 9
Start MooseFS GUI server process if the process is not already running, otherwise display error message.
.IP stop 
Stop MooseFS GUI server process if the process is running, otherwise display error message.
.IP restart 
(default action) If the MooseFS GUI server process is running, restart it, if it's not running, start it. 
.IP reload
Send a signal to MooseFS GUI server process to reload config files.
.IP info
Send a signal to MooseFS GUI server process to dump debug information file in working directory.
.IP test
Displays "mfsgui pid: PID" if the MooseFS GUI server is running (and return 0) or  "mfsgui is not running" otherwise (and return 1).
.IP kill
Send termination signal to MooseFS GUI server process.
.IP restore
Not implemented.
.RE
.SH FILES
.TP
.BR mfsgui.cfg
configuration file for MooseFS GUI server process; refer to
\fBmfsgui.cfg\fP\|(5) manual for details
.TP
.BR .mfsgui.lock
lock file of running MooseFS GUI server process
(created in data directory)
.TP
\fIscripts\fP
various python, javascript, and other (static) files used to serve the GUI interface via HTTP; these files are installed in whatever directory is the default at installation time for \fBROOT_DIR\fP in \fBmfsgui.cfg\fP
.SH "REPORTING BUGS"
Report bugs to <bugs@moosefs.com>.
.SH COPYRIGHT
Copyright (C) 2026 Jakub Kruszona-Zawadzki, Saglabs SA

This file is part of MooseFS.

MooseFS is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, version 2 (only).

MooseFS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, see
<https://www.gnu.org/licenses/>.
.SH "SEE ALSO"
.BR mfsmaster (8),
.BR mfsgui.cfg (5)


NEWS
This file lists noteworthy changes in MooseFS.

* MooseFS 4.59.2-1 (2026-05-14)

  - (cs) fixed bug preventing background chunk testing

* MooseFS 4.59.1-1 (2026-05-04)

  - (cs) fixed potential OOB read

* MooseFS 4.59.0-1 (2026-04-23)

  - (all) changed licence handling (pro only)
  - (all) fixed some security issues

* MooseFS 4.58.5-1 (2026-04-01)

  - (cs) removed aggresive fadvise calls

* MooseFS 4.58.4-1 (2026-03-18)

  - (all) added spec files for Fedora, changed autogen.sh to bootstrap.sh
  - (all) added 'printflike' macro to mfslog, fixed all printflike macros, fixed some syslog args
  - (common) silenced address sanitizer unnecessary warning
  - (common) fixed masterresolver (potential segfault when the last name was empty - pro only)
  - (common) added protection against double initialization of strerr
  - (master) added expected chunk versions to syslog messages

* MooseFS 4.58.3-1 (2025-12-15)

  - (master+tools) fixed conditions for reversible storage class
  - (daemons) fixed small security issue (clearing auxiliary groups before setuid)
  - (libmfsio) fixed descriptor resize function

* MooseFS 4.58.2-1 (2025-11-03)

  - (master) changed timeouts used in master to master connections (pro only)
  - (client) fixed test for overlapping requests in read-ahead algorithm (pull request by Chuandev)
  - (all) fixed typos and spelling errors

* MooseFS 4.58.1-1 (2025-09-29)

  - (master) fix for segfault in flocklocks.c
  - (cli) fixed handling disconnected chunkservers
  - (cli) import fixes in cli mode
  - (tools) improved error messages in mfssetfacl
  - (tools) fixed parsing numerical groups in mfssetfacl
  - (master) prevent setting archive flag for files that are already in trash

* MooseFS 4.58.0-1 (2025-09-15)

  - (master) fixed support for partial readdir
  - (client) added reading dirs in parts
  - (client) added listxattr cache
  - (client) redesigned readdir caches
  - (client) added 'mfsreaddirplusminto' option needed for readdirplus

* MooseFS 4.57.8-1 (2025-08-05)

  - (crc) fixed crc calc for CPU's that can't access memory on non even addresses (i.e. arm-hf)
  - (cs) less disk ops for idle servers
  - (debian) updated postinstall scripts
  - (nbd) fixed restoring nbd devices after sleep/hibernate
  - (cli) fixed displaying master's and chunkserver's charts data
  - (systemd) added start/stop timeouts to chunkserver systemd parameters

* MooseFS 4.57.7-1 (2025-06-16)

  - (all) joined packages cgi and cgiserv to a new gui package
  - (gui) added dedicated GUI server
  - (gui) gui refactor
  - (cli) cli refactor
  - (gui) added support for Prometheus metrics
  - (master) fixed freeing waiting locks from data structures
  - (daemons) added condition to skip socket routines when poll returned zero
  - (daemons) fixed disabling OOM killer
  - (client) removed using potentially dangerous sprintf function

* MooseFS 4.57.6-1 (2025-03-25)

  - (master) fixed predefined storage classes
  - (master) fixed csipmap initialization issue (no csmap after start/restart)
  - (master) fixed lock listing (case of posix_locks and flocks being in the same packet)
  - (master) fixed chunk recovery conditions during read and lookup
  - (master) changed description of default storage classes
  - (man) added manpage for mfsmount.cfg file
  - (man) added subsections to all man pages and reworded some of them

* MooseFS 4.57.5-1 (2025-02-26)

  - (master) added the number of localsplit operations to the total split counter
  - (master) fixed chunk copies and EC matrix counters when both copies and parts are present
  - (master) fixed sending changelog rotate packet to metalogger (CE only)

* MooseFS 4.57.4-1 (2025-02-17)

  - (chunktool) added recursive mode and option to throw away damaged chunks

* MooseFS 4.57.3-1 (2025-02-04)

  - (mfsio) added support for xattr and acl
  - (master) added protection against setting too long xattr
  - (master) fixed case when user changes only grace period in quota

* MooseFS 4.57.2-1 (2025-01-27)

  - (mount) added 'fake' system.posix_acl* xattrs to listxattr (previously not needed)
  - (gui) added cluster traffic charts, updated names
  - (cli) added checks for wrong integer values in options
  - (freebsd) changed port base name in freebsd from moosefs4 to moosefs
  - (master) added KEEP_LEADERSHIP option
  - (gui) charts legend update

* MooseFS 4.57.1-1 (2025-01-08)

  - (master) fixed bug in case of making space for ec parts marked for removal
  - (gui) fixed tmp maintenance time display
  - (gui) fixed graph summary bar display
  - (master) fixed changelog rotation issues
  - (mfsio) fixed mutex initialisation issue
  - (autoconf) updated configure.ac for new PKG_PROG_PKG_CONFIG macro
  - (daemons) added timestamp to info file
  - (master) added meta info to info file
  - (gui) fixed asyncio for python <3.7

* MooseFS 4.57.0-1 (2024-12-09)

  - (all) added storage class priority, description and export group
  - (master) added storage class group allowance to exports.cfg (deprecates mingoal/maxgoal)
  - (master) changed predefined storage classes in new MFS instances
  - (all) removed support for python 2.7
  - (master) always resolve master hostname on info command (PRO only)
  - (master) added to supervisor in 'simple mode' broadcasting LEADER IP to all FOLLOWERs (PRO only)
  - (master) added info for changelog.c
  - (daemons) warnings if a variable is defined more than once in a config file
  - (bdev) added test for posix compliant ioctl prototype
  - (freebsd) moved port files from sysutils to filesystems
  - (master) fixed chunkserver registration in DEPUTY mode (PRO only)
  - (all) added names to some constants used across the system
  - (all) added new charts data (total data I/O from all clients)
  - (metasearch) added printing paths for trash and sustained items
  - (master) fixed constants in metadata.c

* MooseFS 4.56.6-1 (2024-09-23)

  - (cgi+cli) fixed bugs in CLI mode introduced by new CGI features

* MooseFS 4.56.5-1 (2024-09-18)

mfsscripts/mfsgui.py
#!/usr/bin/env python3

# Copyright (C) 2026 Jakub Kruszona-Zawadzki, Saglabs SA
# 
# This file is part of MooseFS.
# 
# MooseFS is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 2 (only).
# 
# MooseFS is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, see
# <https://www.gnu.org/licenses/>.

import sys

if sys.version_info[0]<3 or (sys.version_info[0]==3 and sys.version_info[1]<4):
	print("Content-Type: text/html; charset=UTF-8\r\n\r")
	print("""<!DOCTYPE html>""")
	print("""<html lang="en">""")
	print("""<head>""")
	print("""<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />""")
	print("""<title>MooseFS</title>""")
	print("""</head>""")
	print("""<body>""")
	print("""<h2>MooseFS GUI: Unsupported python version, minimum required version is 3.4</h2>""")
	print("""</body>""")
	print("""</html>""")
	sys.exit(1)

from common.constants_ac import * # MFS_INLINE_IMPORT
from common.constants import *    # MFS_INLINE_IMPORT
from common.errors import *       # MFS_INLINE_IMPORT
from common.utils import *        # MFS_INLINE_IMPORT
from common.conn import *         # MFS_INLINE_IMPORT
from common.models import *       # MFS_INLINE_IMPORT
from common.cluster import *      # MFS_INLINE_IMPORT
from common.dataprovider import * # MFS_INLINE_IMPORT
from common.organization import * # MFS_INLINE_IMPORT
from common.utilsgui import *
import common.validator as validator

ajax_request = AJAX_NONE #is it ajax request?
readonly = False     #if readonly - don't render CGI links for commands
selectable = True    #if not selectable - don't render drop-downs and other selectors (to switch views with page reload)
donotresolve = 0     #resolve or not various ip addresses
instancename = "My MooseFS"

masterhost = DEFAULT_MASTERNAME
masterport = DEFAULT_MASTER_CLIENT_PORT



# set default output encoding to utf-8 (our html page encoding)
if sys.version_info[1]<7:
	import codecs
	sys.stdout = codecs.getwriter("utf-8")(sys.stdout.detach())
else:
	sys.stdout.reconfigure(encoding='utf-8')

fields = MFSFieldStorage()
fields.initFromURL()		

if fields.getstr("ajax")=="container":
	ajax_request = AJAX_CONTAINER
elif fields.getstr("ajax")=="metrics":
	ajax_request = AJAX_METRICS

if fields.getvalue("ajax")!=None:
	fields.pop("ajax") #prevent from including ajax URL param any further

try:
	if "masterhost" in fields:
		masterhost = fields.getstr("masterhost")
		if type(masterhost) is list:
			masterhost = ";".join(masterhost)
		masterhost = masterhost.replace('"','').replace('<','').replace('>','').replace("'",'').replace('&','').replace('%','')
except Exception:
	pass
masterport = int(fields.getint("masterport",DEFAULT_MASTER_CLIENT_PORT))
mastername = fields.getstr("mastername", "")

# connect to masters and find who is the leader etc
cl = Cluster(masterhost, masterport, fields.getstr("leaderip"))
if (cl.leaderfound()):
	fields.append("leaderip",cl.master().host)


html_title = "MooseFS"
if len(mastername)>0: html_title += " (%s)" % htmlentities(mastername)

errmsg = None
if cl.master()==None:
	errmsg = cl.errormsg()
if (cl.leaderfound() or cl.electfound() or cl.usurperfound() or cl.followerfound()):
	if cl.master().version_unknown():
		errmsg = """Can't detect the MooseFS Master server version (%s)""" % (masterhost)
	elif cl.master().version_less_than(3,0,0):
		errmsg = """This version of MooseFS Master server (%s) is not supported (pre 3.0.0)""" % (masterhost)
if errmsg:
	print("Content-Type: text/html; charset=UTF-8\r\n\r")
	print("""<!DOCTYPE html>""")
	print("""<html lang="en">""")
	print("""<head>""")
	print("""<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />""")
	print("""<title>MooseFS GUI %s</title>""" % (htmlentities(mastername)))
	# leave this script a the beginning to prevent screen blinking when using dark mode
	print("""<script type="text/javascript"><!--//--><![CDATA[//><!--
		if (localStorage.getItem('theme')===null || localStorage.getItem('theme')==='dark') { document.documentElement.setAttribute('data-theme', 'dark');}	
		//--><!]]></script>""")		
	print("""<link rel="stylesheet" href="assets/mfs.css" type="text/css" />""")
	print("""</head>""")
	print("""<body>""")
	print("""<h1 class="center">%s</h1>""" % htmlentities(errmsg))
	vstr = VERSION
	print("""<h4 class="center">GUI&nbsp;v.%s&nbsp;&nbsp;Python v.%u.%u</h4>""" % (vstr, sys.version_info[0], sys.version_info[1]))
	print("""</body>""")
	print("""</html>""")
	sys.exit(1)

# initialize global data provider
dataprovider = DataProvider(cl, donotresolve)
# initialize global validator
vld = validator.Validator(dataprovider)

# prior to rendering anything process commands if any
import views.commands
views.commands.process_commands(cl,fields,html_title)

# organization of menus, sections and subsections
org = Organization(dataprovider,guimode=True, ajax_request=ajax_request)

# decode URL parameters
if "readonly" in fields: #readonly=1 - don't show any command links (on/off maintenance, remove etc.)
	if fields.getvalue("readonly")=="1":
		readonly = True
if "selectable" in fields: #selectable=0 - don't show any selectors requiring page refresh (knobs, drop-downs etc.)
	if fields.getvalue("selectable")=="0":
		selectable = False
if "sections" in fields: #list of "-" separated sections to show
	sectionstr = fields.getvalue("sections")
	if type(sectionstr) is list:
		sectionstr = "-".join(sectionstr)
	sectionset = set(sectionstr.split("-"))
else:
	sectionset = set((org.get_default_section(),)) # by default:show the 1st section if no section is defined in url
if "subsections" in fields: #list of "-" separated subsections to show
	subsectionstr = fields.getvalue("subsections")
	if type(subsectionstr) is list:
		subsectionstr = "-".join(subsectionstr)
	subsectionset = set(subsectionstr.split("-"))
else:
	subsectionset = set([])

org.set_sections(sectionset,subsectionset)

#######################################################
###                 RENDER OUTPUT                   ###
#######################################################

# Print Prometheus metrics and exit
if ajax_request == AJAX_METRICS: 
	import views.metrics as metrics
	metrics.print_render(fields, masterhost, masterport)
	exit(0)

# print header
print("Content-Type: text/html; charset=UTF-8\r\n\r")

# print html header and menu
if not ajax_request:
	import views.header
	print_out(views.header.render(dataprovider, fields, vld, html_title, org, sectionset))
	print("""<div id="container">""")
	print("""<div id="container-ajax">""")
	if cl.leaderfound()==0:
		# deal with missing leader in the cluster
		import views.missingleader
		print_out(views.missingleader.render(dataprovider, fields, vld))



#######################################################
###                 RENDER SECTIONS                 ###
#######################################################

# Status section
if org.shall_render("ST"):
	try:
		import views.graph
		print_out(views.graph.render(dataprovider, fields, vld))
	except Exception:
		print_exception()	

# Info section

#Cluster summary subsection
if org.shall_render("IG"):
	try:
		ci = dataprovider.get_clusterinfo()
		import views.clusterinfo
		print_out(views.clusterinfo.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

#Metadata servers subsection
if org.shall_render("IM"):
	try:		
		# update master servers delay times prior to getting the list of master servers
		highest_saved_metaversion, highest_metaversion_checksum = cl.update_masterservers_delays()
		import views.metaservers
		print_out(views.metaservers.render(dataprovider,fields,vld))
	except Exception:
		print_exception()


# Chunks matrix
if org.shall_render("IC"):
	try:
		import views.matrix 
		print_out(views.matrix.render(dataprovider, fields, vld, selectable))
	except Exception:
		print_exception()

# Filesystem self-check loop (GUI only)
if org.shall_render("FL"):
	try:
		import views.healthselfcheck
		print_out(views.healthselfcheck.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

# Chunk house-keeping loop (GUI only)
if org.shall_render("CL"):
	try:
		import views.chunksloop
		print_out(views.chunksloop.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

# memory usage table
if org.shall_render("MU"):
	try:
		# mu.memlabels,abrlabels,mu.totalused,mu.totalallocated,memusage = dataprovider.get_memory_usage()
		mu = dataprovider.get_memory_usage()
		import views.memory
		print_out(views.memory.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

# Chunkservers section
if org.shall_render("CS"):
	try:
		import views.chunkservers 
		print_out(views.chunkservers.render(dataprovider, fields, vld, readonly, selectable))
	except Exception:
		print_exception()

# Disks section
if org.shall_render("HD"):
	try:
		import views.disks
		print_out(views.disks.render(dataprovider, fields, vld, readonly, selectable))
	except TimeoutError:
		print_error("Timeout connecting chunk servers. Hint: check if chunk servers are available from this server (the one hosting GUI/CLI).")
	except Exception:
		print_exception()

# Exports section
if org.shall_render("EX"):
	try:
		import views.exports
		print_out(views.exports.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

if org.shall_render("MD"):
	html_menu_subsections(org, fields, "MD")

# Mounts - parameters section
if org.shall_render("MS"):
	try:
		import views.mountsparameters
		print_out(views.mountsparameters.render(dataprovider, fields, vld, readonly))
	except Exception:
		print_exception()

# Mounts - operations section
if org.shall_render("MO"):
		try:
			import views.mountsoperations
			print_out(views.mountsoperations.render(dataprovider, fields, vld))
		except Exception:
			print_exception()

# Open files section
if org.shall_render("OF"):
	try:
		import views.openfiles
		print_out(views.openfiles.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

if org.shall_render("RP"):
	html_menu_subsections(org, fields, "RP")
	
# Storage classes section
if org.shall_render("SC"):
	try:
		import views.sclasses
		print_out(views.sclasses.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

# Override patterns section
if org.shall_render("PA"):
	try:
		import views.opatterns
		print_out(views.opatterns.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

# Quotas
if org.shall_render("QU"):
	try:
		import views.quotas
		print_out(views.quotas.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

if not ajax_request:
	print("""</div><!-- end of container-ajax -->""")
	# content from now on (i.e., charts) is not subject of the full DOM ajax update

# Master and Chunkservers charts
if org.shall_render("XC"):
	html_menu_subsections(org, fields, "XC")

# Include servers charts scripts
if (org.shall_render("MC") or org.shall_render("CC")):
	print("""<script src="assets/acidchart.js" type="text/javascript"></script>""")

# Master servers charts
# charts are ajax-updated individually, without resending the page content
if org.shall_render("MC"):
	out = []
	try:
		import views.chartsmc
		print_out(views.chartsmc.render(dataprovider, fields, vld))
	except Exception:
		print_exception()

# Chunkservers charts
# charts are ajax-updated individually, without resending the page content
if org.shall_render("CC"):
		out = []
		try:
			import views.chartscc
			print_out(views.chartscc.render(dataprovider, fields, vld))
		except Exception:
			print_exception()


#######################################################
###                 RENDER FOOTER                   ###
#######################################################

if not ajax_request:
	print("""</div><!-- end of container -->""")
	import views.footer
	print_out(views.footer.render(dataprovider, fields, vld))
